python/tutorials/02-fused-softmax.py

```diff
@@ -71,58 +72,85 @@ def naive_softmax(x):
 
 
 @triton.jit
-def softmax_kernel(output_ptr, input_ptr, input_row_stride, output_row_stride, n_cols, BLOCK_SIZE: tl.constexpr):
-    # The rows of the softmax are independent, so we parallelize across those
-    row_idx = tl.program_id(0)
-    # The stride represents how much we need to increase the pointer to advance 1 row
-    row_start_ptr = input_ptr + row_idx * input_row_stride
-    # The block size is the next power of two greater than n_cols, so we can fit each
-    # row in a single block
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    input_ptrs = row_start_ptr + col_offsets
-    # Load the row into SRAM, using a mask since BLOCK_SIZE may be > than n_cols
-    row = tl.load(input_ptrs, mask=col_offsets < n_cols, other=-float('inf'))
-    # Subtract maximum for numerical stability
-    row_minus_max = row - tl.max(row, axis=0)
-    # Note that exponentiation in Triton is fast but approximate (i.e., think __expf in CUDA)
-    numerator = tl.exp(row_minus_max)
-    denominator = tl.sum(numerator, axis=0)
-    softmax_output = numerator / denominator
-    # Write back output to DRAM
-    output_row_start_ptr = output_ptr + row_idx * output_row_stride
-    output_ptrs = output_row_start_ptr + col_offsets
-    tl.store(output_ptrs, softmax_output, mask=col_offsets < n_cols)
+def softmax_kernel(output_ptr, input_ptr, input_row_stride, output_row_stride, n_rows, n_cols, XBLOCK: tl.constexpr,
+                   RBLOCK: tl.constexpr, NUM_REMAINDER_ROWS: tl.constexpr, num_stages: tl.constexpr):
+    pid = tl.program_id(0)
+    # starting row of the program
+    row_start = pid * (XBLOCK - 1) + min(pid, NUM_REMAINDER_ROWS)
```

Comment:
I think you could simplify this and get rid of `XBLOCK` and `NUM_REMAINDER_ROWS` by using a sort of grid-stride loop:
```
row_start = tl.program_id(0)
row_step = tl.num_programs(0)
for i in tl.range(row_start, n_rows, row_step, num_stages=num_stages):
    ...
 ```

Reply:
This is definitely simpler than the current implementation! I thought having a program reference consecutive rows during pipelining could help L2-cache performance. But from the results the simpler implementation is better in most of the cases, expect for the last one. Maybe we should just go with the simplification.

Stage=4, num_warps = 8:
         ```
softmax-performance:
          N       Triton  Torch (native)  Torch (jit)
0     256.0   414.784823      824.352186   318.909986
1     384.0   601.247730     1043.013292   412.176106
2     512.0   791.975859     1213.629628   481.439841
3     640.0   925.649747     1340.204504   526.816702
4     768.0  1077.304067     1448.309387   565.371676
..      ...          ...             ...          ...
93  12160.0  2034.780648     1664.461988   617.957313
94  12288.0  2022.000925     1616.821281   619.374961
95  12416.0  2039.949273     1603.478860   618.835915
96  12544.0  2026.353660     1604.829503   620.204555
97  12672.0  1957.183654     1591.577105   619.563046
```